Add Spark Job

holdenk opened the pull request:
Initial work in progress attempt at adding Spark to kubeflow using the spark-on-k8s-operator as a base starting point. This is super early but I'd love peoples feedback on the direction with this.

cc @texasmichelle 

Known TODOs:
- Tests that run on more than just my specific directory layout
- Documentation
- Cleanup

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kubeflow/1467)
<!-- Reviewable:end -->

pdmack commented on kubeflow/spark/parts.yaml:
```diff
@@ -0,0 +1,27 @@
+   "name": "spark",
+   "apiVersion": "0.0.1",
+   "kind": "ksonnet.io/parts",
+   "description": "An empty package used as a stub for new packages.\n",
```
"Holden's awesome Spark Job prototype\n"

holdenk replied:
<3 :D

pdmack commented:
/ok-to-test

@holdenk did you autoformat the *sonnet files yet? Guess we'll find out...

pdmack commented:
winner, winner...

Please use `scripts/autoformat_jsonnet.sh` to fix up the sonnet files.

inc0 commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,339 @@
+            },
+            labels: {
+              "app.kubernetes.io/name": name + "-sparkoperator",
+              "app.kubernetes.io/version": "v2.3.1-v1alpha1",
```
should this be a parameter with default ver?

holdenk replied:
Yes, good catch,

inc0 commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,339 @@
+        {
+          kind: "ServiceAccount",
+          name: name + "-spark",
+          namespace: "default",
```
this should be param

holdenk replied:
thanks for catching.

holdenk commented:
Hey @texasmichelle & @kunmingg I'm wondering about the best way to test this operator - looking at the `test_deploy.py` script it seems to use KF components @ master but since this is a PR I'm not sure that would work well.

kunmingg commented:
@holdenk 
1. You can edit tmp logic change in ```test_deploy.py``` like changing version to use.
2. presubmit test will then test under your new logic.
3. When test looks good you can revert change in 1. and merge this PR.

holdenk commented:
Ok so I can point it at this branch. Is there a reason this isn't auto
pointed at the branch in PR builds (seems like it could give us bad results
on PRs in general)? Unless I'm missing something which is quite possible
since it's my first pass through this code.

On Fri, Oct 19, 2018, 3:43 PM Kunming Qu <notifications@github.com> wrote:

> @holdenk <https://github.com/holdenk>
>
>    1. You can edit tmp logic change in test_deploy.py like changing
>    version to use.
>    2. presubmit test will then test under your new logic.
>    3. When test looks good you can revert change in 1. and merge this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/kubeflow/pull/1467#issuecomment-431519685>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADp9Yti_u4x1R_FDkw4E8_59xz5UbZZks5umlWugaJpZM4WX-c_>
> .
>

kunmingg commented:
@holdenk 
For tests running on GKE I believe now they are covered by https://github.com/kubeflow/kubeflow/blob/master/prow_config.yaml#L15
which should deploy components from current PR code.
It would be easier to add new test here.

```test_deploy.py``` is more focusing on minikube integration now.

kris-nova commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,339 @@
+      },
+    },
+    operatorClusterRole:: {
+      kind: "ClusterRole",
```
Why `ClusterRole` instead of `Role`? Looks like we are binding on a single namespace only, and opening up the broader permissions might be unnecessary? 

https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole

holdenk replied:
Not super sure since the RBACS are based on the ones from the spark-operator project from GCP folks, my gut is that if we wanted to support having jobs in different namesapces wed need the operator have a `ClusterRole` but I'm not super sure. The RBAC file for the spark-operator is at https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/8c7fdbb306dfd656093c1b2a4ede901d651c9bd5/manifest/spark-operator-rbac.yaml , I could try and scope it down though and see if it still works?

kris-nova replied:
Ah I didn't realize it was a port from the `spark-operator` project. I can ask there. It's not necessarily a concern, just more of wandering why we needed the broader scope. If we are spanning namespaces that makes sense. No change needed, thanks for clarifying.

holdenk commented:
/retest

holdenk commented:
/retest

holdenk commented:
/retest

jlewi commented on testing/deploy_kubeflow.py:
```diff
@@ -56,6 +57,8 @@ def deploy_kubeflow(test_case):
   # TODO(jlewi): We don't need to generate a core component if we are
   # just deploying TFServing. Might be better to refactor this code.
   # Deploy Kubeflow
+  print("Generate operators.")
```
Is this script called by kfctl_test.jsonnet?

I think kfctl_test.jsonnet is just calling kfctl.sh and not using these script..

holdenk replied:
No, this is called by the e2e workflow test.

jlewi commented on testing/workflows/components/kfctl_test.jsonnet:
```diff
@@ -215,6 +215,11 @@ local dagTemplates = [
+        // Generate spark operator
+        "&&",
+        kfCtlPath,
+        "generate",
```
Did you modify kfctl to add this?

jlewi commented on components/k8s-model-server/images/releaser/components/workflows.libsonnet:
```diff
@@ -247,6 +247,11 @@
           dependencies: ["deploy-tf-serving"],
         },
+        {
+          name: "deploy-spark-job",
```
Why did you modify this file?
I think this E2E test is specific for K8s model server.

jlewi commented:
It doesn't look like the test is quite right. 

Here's some guidance.

You want to create an Argo Workflow that does something like the following

1. Use kfctl.sh to create and deploy Kubeflow
2. Run the ksonnet command to generate the spark job component and apply it
3. Run some script e.g. the python script you have to verify the deployment is working.

For now I don't think we should install the sparkjob operator by default; so for now users should have to do something like

ks generate spark-job spark-job
ks apply ${ENV} -c spark-job

So your Argo workflow should have steps executing these commands.

In terms of which Argo workflow to modify I think you want to look at this one
https://github.com/kubeflow/kubeflow/blob/master/testing/workflows/components/kfctl_test.jsonnet

You have 2 options

1. You can add an optional step to that Argo workflow to deploy/test the spark job
1. You could create a separate Argo workflow to just test deploy the SparkJob CRD.

holdenk commented on testing/deploy_utils.py:
```diff
@@ -97,14 +97,22 @@ def setup_kubeflow_ks_app(dir, namespace, github_token, api_client):
     cwd=app_dir)
 
-  kubeflow_registry = "github.com/kubeflow/kubeflow/tree/master/kubeflow"
+  # IDK what the env looks like lets just dump everything
```
This hard coded registry to master was confusing to me, and using the one that's checked out seems to work -- is there a reason for the master registry?

jlewi replied:
Why are you modifying this script?

jlewi replied:
Are you using this script test_deploy.py to deploy the spark operator in your cluster?

holdenk replied:
So I figured that hard coding master was a bad idea and installing Spark as part of the e2e minikube tests make sure it at least can be installed. I don't feel strongly about this since we don't need it for the full workflow tests so I'm happy to revert if this complicates matters.

jlewi commented:
In terms of getting the logs. If you add labels to the argo steps as described here
https://github.com/kubeflow/testing#argo-spec

That should make it more convenient to get pod ids.

Unfortunately most of our existing workflow tests e.g. kfctl_test.jsonnet don't do this so you'd have to add it.

Assuming you have the right access (I think you added yourself to ci-team?) you should be able to access the kubeflow-testing cluster so you should be able to then get the pod id from the labels.

Instructions for fetching the logs are here:
https://github.com/kubeflow/testing#debugging-failed-tests

As an example here's a query to fetch one of the pod logs

```
resource.type="container"
resource.labels.cluster_name="kubeflow-testing"
resource.labels.pod_id="kubeflow-tf-operator-presubmit-v1beta1-910-ec293c3-1353-4037-3768977243"
```

holdenk commented:
Ok, I'd been trying to find the pod ids in the gubernator logs as mentioned in https://github.com/kubeflow/testing#debugging-failed-tests but it wasn't really working out ( see https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/kubeflow_kubeflow/1467/kubeflow-presubmit/4917/?log#log ). Connecting to the kubeflow-testing cluster though I see only pods > 10 days old which doesn't seem right.

holdenk commented:
It sounds like it might be easier to test as a separate workflow so I guess I'll explore how to do that assuming that `deploy_kubeflow.py` isn't meant to be general purpose.

holdenk commented:
Looks like the E2E tests finally pass, I'll rework the commit history into something more manageable but if you've got some cycles to take a look at how it ended up would be rad cc @texasmichelle

jlewi commented on testing/spark_temp/simple_test.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
What is this script for? Can you add a comment please?

holdenk replied:
This was for local testing, I can remove it.

jlewi commented on testing/workflows/components/kfctl_test.jsonnet:
```diff
@@ -230,6 +230,53 @@ local dagTemplates = [
   },
+  {
+    template: buildTemplate(
+      "install-spark-operator",
```
So here in your test you are running ks directly to install the package so are you using deploy_kubeflow.py or test_deploy.py?

holdenk replied:
Yeah I can revert those if we want, I figured it made sense to see Spark installed on minikube even if we only used the operator on the full version in e2e workflow.

jlewi commented on testing/workflows/components/workflows.libsonnet:
```diff
@@ -278,6 +278,25 @@
         },
         dependencies: ["wait-for-kubeflow"],
       },  // pytorchjob - deploy,
+      {
```
I think this is legacy; I think this workflow is only used by the minikube test because it hasn't been updated to use kfctl.sh. 

I think its fine not to run the test on minikube.

holdenk replied:
So this part seems to be triggered outside of the minikube tests.

That being said, I think it might make sense, in the future, to test this and other operator one-by-one on minikube, what do you think?

jlewi commented:
Really appreciate your persistence; I know this has been a slog through our test infra.

I think you are making some unnecessary changes related to testing.

Some background:

There are currently two different E2E tests for deploying Kubeflow and verifying it works

1. [kfctl_test.jsonnet](https://github.com/kubeflow/kubeflow/blob/master/testing/workflows/components/kfctl_test.jsonnet) - Uses GKE and kfctl.sh
1. [workflows.jsonnet](https://github.com/kubeflow/kubeflow/blob/master/testing/workflows/components/workflows.jsonnet) - Uses minikube and kfctl.sh

Both of these use the code in workflows.libsonnet.

The minikube way of doing things is outdated as described in this TODO
https://github.com/kubeflow/kubeflow/blob/master/testing/workflows/components/workflows.libsonnet#L325

We want to follow the GKE test and use kfctl.sh to deploy things.

As result, you should only test on GKE not on minikube because as the above indicates the minikube test infra is not well maintained.

This means you shouldn't need to modify deploy_kubeflow.py.

So the next thing you need to figure out is how will you add steps to the argo workflow to install the spark job package and verify it works. There are two ways to do this

1. You can add argo steps that just shell out to ks
    * This appears to be what you are doing in kfctl_test.sh https://github.com/kubeflow/kubeflow/pull/1467/files#diff-4718be91c949a9485ca0cf3fa589267aR234

1. You can create a script (bash or python) to run the commands

    * It seems like you are partially going down this road by creating simple_test.sh but it doesn't seem like this script gets used anywhere.

My suggestion would be to go the route of the second option and use python. The main advantage is that you can then run the test locally against your K8s cluster to see if its working.

For an example of what this python script might look like you can take a look at

https://github.com/kubeflow/examples/blob/master/mnist/testing/deploy_test.py

   * This script runs some ks commands to deploy a TFServing instance
   * Runs a simple check to verify that its deployed successfully
   * You'd have to adapt the commands of course

The main advantage is that as the instructions in that file show; it should be fairly straightforward to run it against your existing K8s cluster to develop and debug.

Once you have the script working you can just invoke it from your Argo workflow. This way integration into Kubeflow CI is largely just a matter of getting the paths, working directory, etc... set correctly.

holdenk commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,348 @@
+    // generally want to use that namespace for a component.
+    // However, in some cases an application may use multiple namespaces in which
+    // case the namespace for a particular component will be a parameter.
+    local namespace = if std.objectHas(params, "namespace") then params.namespace else env.namespace,
```
Chatted with @texasmichelle about this and the weird behaviour I had scene back with the minikube tests and it make sense now so I'll keep as-is

holdenk commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,348 @@
+          instances: 1,
+          memory: params.executorMemory,
+          labels: {
+            version: "2.3.1",
```
Make the version come from params?

holdenk commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,348 @@
+          ],
+        },
+        executor: {
+          cores: 1,
```
TODO: paramatirize this

holdenk commented on testing/deploy_kubeflow.py:
```diff
@@ -114,6 +114,7 @@ def deploy_kubeflow(test_case):
 
 
 def main():
+  print("Calling deploy kubeflow")
```
REVERT THIS

holdenk commented on testing/test_deploy.py:
```diff
@@ -126,7 +126,7 @@ def setup_kubeflow_ks_app(args, api_client):
 
   # Install required packages
-  packages = ["kubeflow/common", "kubeflow/tf-serving", "kubeflow/tf-job", "kubeflow/pytorch-job", "kubeflow/argo"]
+  packages = ["kubeflow/common", "kubeflow/tf-serving", "kubeflow/tf-job", "kubeflow/pytorch-job", "kubeflow/argo", "kubeflow/spark"]
```
TODO: revert this

jlewi commented:
@holdenk Anything I can do to help?

holdenk commented on testing/deploy_utils.py:
```diff
@@ -97,14 +97,22 @@ def setup_kubeflow_ks_app(dir, namespace, github_token, api_client):
+  packages = ["kubeflow/common", "kubeflow/gcp", "kubeflow/jupyter", "kubeflow/tf-serving", "kubeflow/tf-job", "kubeflow/tf-training", "kubeflow/pytorch-job", "kubeflow/argo", "kubeflow/spark"]
 
   # Instead of installing packages we edit the app.yaml file directly
+  # ^ Why?
```
I wish I knew why, but it is unrelated so I'll get rid of this.

holdenk commented on kubeflow/spark/all.libsonnet:
```diff
@@ -0,0 +1,339 @@
+            "nodes",
+          ],
+          verbs: [
+            "get",
```
So ClusterRole versus Role is now user configurable, if folks don't need to run jobs outside of the namespace where they created the operator we'll just do a Role, but if they want to have the operator and jobs sit in different namespaces we use clusterrole.

holdenk commented:
Thanks so much for all your help with reviews @jlewi , @texasmichelle  @inc0 , @pdmack, @kris-nova & @kunmingg it should be ready for review again I've squashed it down and rebased on master.

jlewi commented:
FYI looks like the test is failing

```
+ [[ 1 -eq 0 ]]
+ echo Command failed: ks apply -c spark-operator --verbose
+ for i in '$(seq 1 3)'
+ set +e
+ ks apply -c spark-operator --verbose
level=debug msg="setting log verbosity" verbosity-level=1
level=debug msg="loading application configuration from /mnt/test-data-volume/kubeflow-presubmit-kfctl-1467-c45c69c-5289-6200/kfctl-6200/ks_app"
level=error msg="environment is not set; use `env list` to see available environments"
```

holdenk commented:
/test kubeflow-presubmit

holdenk commented:
cc @jlewi can you tag this as OK to test again? CI lost the tag :(

richardsliu commented:
/ok-to-test

holdenk commented:
Thanks @richardsliu !

holdenk commented:
/test kubeflow-presubmit

holdenk commented:
Looks like the spark operator is applying successfully now, so it should be ready for review again. cc @jlewi @texasmichelle

jlewi commented:
Thanks @holdenk. It looks like the spark apply test is still failing

```
+ ks apply default -c spark-operator --verbose
level=debug msg="setting log verbosity" verbosity-level=1
level=debug msg="loading application configuration from /mnt/test-data-volume/kubeflow-presubmit-kfctl-1467-39b503c-5575-5aca/kfctl-5aca/ks_app"
level=debug msg="creating ks pipeline for environment \"default\""
level=debug msg="building objects" action=pipeline module-name=/
level=debug msg="jsonnet evaluate snippet" elapsed=59.03316ms name=applyGlobals
level=error msg="find objects: open /mnt/test-data-volume/kubeflow-presubmit-kfctl-1467-39b503c-5575-5aca/kfctl-5aca/ks_app/environments/default/params.li
bsonnet: no such file or directory"
```

I think the problem is you aren't creating the default environment.
The default environment is created by the kfctl-apply-k8s step. So I think you want to make the spark-operator step depend on that step.

jlewi commented:
/ok-to-test

holdenk commented:
@jlewi so I think the apply operator is succeeding, the part which is failing is the part which depended on the Python script so I don't think it the env issue (although that was possibly the issue before).

jlewi commented:
Woo Hoo!
Thanks for pushing this through.
/lgtm
/approve

holdenk commented:
Test failures look unrelated, /retest

jlewi commented:
We were having quota issues earlier. Should be fixed now.

/lgtm
/approve

holdenk commented:
It's still failing before getting to the Spark specific code, and looks like a quota failure so lets try:
/retest

holdenk commented:
Although IDK if the bot listens to me might need someone else to tell it to retest

holdenk commented:
Ok the spark jobs passed but the notebooks test failed ?

holdenk commented:
/retest

holdenk commented:
/meow
